### ISSUE TITLE
twisted.words.im.ircsupport.IRCProto sends USER command with the wrong data

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4641]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4641|
|Type|release blocker: regression|
|Created|2010-08-26 02:49:48Z|
|Branch|https://github.com/twisted/twisted/tree/ircsupport-user-4641|

When `IRCProto.connectionMade` is called, a `NICK` command is first sent using `IRCClient.setNick`.  Then a `USER` command is sent using `self.nickname`.  Prior to [#3377](https://github.com/twisted/twisted/issues/3377) (r27307), `setNick` synchronously set the `nickname` attribute to the value passed to it.  Now, however, the attribute is only set after the server acknowledges the `NICK` command.  This causes `IRCProto` to send a `USER` command with a bogus value (it happens to be `"irc"`, as that is the default value for the `nickname` attribute).

<details><summary>Searchable metadata</summary>

```
trac-id__4641 4641
type__release_blocker__regression release blocker: regression
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_ircsupport_user_4641 branches/ircsupport-user-4641
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1282790988000000 1282790988000000
changetime__1284121245000000 1284121245000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [29945]) Branching to 'ircsupport-user-4641'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/ircsupport-user-4641)

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_3"></a>|@jml set owner to @exarkun|
|-|-|

Nice. Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set status to `closed`|
|-|-|

(In [29993]) Merge ircsupport-user-4641

Author: exarkun
Reviewer: jml
Fixes: [#4641](https://github.com/twisted/twisted/issues/4641)

Fix a bug in `twisted.words.im.ircsupport`, introduced by [#3377](https://github.com/twisted/twisted/issues/3377), which caused
the wrong username to be sent at login time.  Also add a few basic tests
covering both this and a couple other login-time behaviors.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
